test/rules/dead-type-branch/function-type-guards/test.ts.lint

```diff
@@ -0,0 +1,40 @@
+class A {private a(){}};  // implicitly declares a type, and methods for
+class B {private b(){}};  // incompatibility. This behaves weirdly when types
+class C {private c(){}};  // are compatible.
+
+const ab: A|B = Math.random() > .5 ? new A() : new B();
+
+function isA(x:any): x is A { return Math.random() > .5; }
+function isB(x:any): x is B { return  Math.random() > .5; }
+function isC(x:any): x is C { return  Math.random() > .5; }
+
+
+if (isA(ab)) { ab; }
+
+
+if (isA(ab) || isB(ab)) { ab; }
+
+
+if (isC(ab)) { ab; } // Could be A & C | B & C
+
+
+if (isA(ab) || isB(ab)) { ab; }
+else { ab; }
+     ~~~~~~~ [This block is unreachable: TypeScript infers the type of 'ab' as 'never' at this location, which means the previous type guard is always false.]
+
+
+// Interesting case: the first condition leaves B & !A, which gets rounded
+// up into B for the next condition since TS doesn't subtract. We lost the !A
+// along the way, which then gets turned into B & !A, and so on.
+if (isA(ab)) { ab; }
+else if (isA(ab)) { ab; }
+else if (isA(ab)) { ab; }
+
+
+if (isA(ab)) { ab; }
+else if (isB(ab)) { ab; }
+else if (isC(ab)) { ab; }
+                  ~~~~~~~ [This block is unreachable: TypeScript infers the type of 'ab' as 'never' at this location, which means the previous type guard is always false.]
+else { ab; }
+     ~~~~~~~ [This block is unreachable: TypeScript infers the type of 'ab' as 'never' at this location, which means the previous type guard is always false.]
+
```

Comment:
consider the following:
```ts
function foo(v: 1 | 2) {
    switch (v) {
        case 1:
        case 2:
            return v;
        default:
            doSomethingWithoutV();
    }
}
```
If I'm not mistaken, the rule will not show a warning on the default clause because it doesn't use `v`.
IIRC there's no way to retrieve the type of a symbol at a location where that symbol is not used. That means the TypeChecker cannot tell you that `v` is `never` because `v` is not used in the default clause.

That said, it seems like this rule only prevents unintentional usage of a variable which type is narrowed to `never`.
I guess that could be achieved without control flow analysis.

Reply:
You're right: if you're not using the value in a block or in its parents, nothing will warn. But it seems risky to rely on that value not being used in that pattern, no ?

(btw, the rule currently doesn't touch case statements at all, but the idea is the same)